### PR TITLE
{base}[GCC/10.3.0] Kokkos v3.4.01

### PR DESCRIPTION
**Pull request by robert-mijakovic**
(created using `eb --new-pr`)

### CODE REVIEW

File: easybuild/easyconfigs/k/kokkos/kokkos-3.4.01-GCC-10.3.0.eb [hunk 1]
```diff
@@ -0,0 +1,43 @@
+##
+# Author:    Robert Mijakovic <robert.mijakovic@lxp.lu>
+##
+easyblock = 'CMakeMake'
+
+name = 'kokkos'
```
Comment (boegel) [inline easybuild/easyconfigs/k/kokkos/kokkos-3.4.01-GCC-10.3.0.eb:6]: @robert-mijakovic Should be `Kokkos`? See https://kokkos.org/about/

File: easybuild/easyconfigs/k/kokkos/kokkos-3.4.01-GCC-10.3.0.eb [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+version = '3.4.01'
+
+homepage = 'https://kokkos.org'
+description = """Kokkos Core implements a programming model in C++ for writing
+                 performance portable applications targeting all major HPC
```
Comment (boegel) [inline easybuild/easyconfigs/k/kokkos/kokkos-3.4.01-GCC-10.3.0.eb:11]: Please strip leading spaces to a single one, or the output of `module show` is going to look weird...

File: easybuild/easyconfigs/k/kokkos/kokkos-3.4.01-GCC-10.3.0.eb [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+source_urls = ['https://github.com/%(name)s/%(name)s/archive/refs/tags/']
+sources = ['%(version)s.tar.gz']
+checksums = ['146d5e233228e75ef59ca497e8f5872d9b272cb93e8e9cdfe05ad34a23f483d1']
+
+builddependencies = [
+    ('CMake', '3.20.1'),
+]
+
+dependencies = [
+    ('binutils', '2.36.1'),
```
Comment (boegel) [inline easybuild/easyconfigs/k/kokkos/kokkos-3.4.01-GCC-10.3.0.eb:30]: No need for the binutils dependency when using `GCC` as toolchain?

File: easybuild/easyconfigs/k/kokkos/kokkos-3.4.01-GCC-10.3.0.eb [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    ('hwloc', '2.4.1'),
+]
+
+configopts = '-DKokkos_ENABLE_OPENMP=On'
+configopts += '-DKokkos_ARCH_HSW=On'
+configopts += '-DKokkos_ENABLE_HWLOC=On'
+configopts += '-DKokkos_HWLOC_DIR=$EBROOTHWLOC'
+
+sanity_check_paths = {
+    'files': ['bin/hpcbind', 'bin/kokkos_launch_compiler', 'bin/nvcc_wrapper'] +
```
Comment (boegel) [inline easybuild/easyconfigs/k/kokkos/kokkos-3.4.01-GCC-10.3.0.eb:40]: Any sensible `sanity_check_commands` that can be added?

File: easybuild/easyconfigs/k/kokkos/kokkos-3.4.01-GCC-10.3.0.eb [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+    ('CMake', '3.20.1'),
+]
+
+dependencies = [
+    ('binutils', '2.36.1'),
+    ('hwloc', '2.4.1'),
+]
+
+configopts = '-DKokkos_ENABLE_OPENMP=On'
+configopts += '-DKokkos_ARCH_HSW=On'
```
Comment (boegel) [inline easybuild/easyconfigs/k/kokkos/kokkos-3.4.01-GCC-10.3.0.eb:35]: What does the `HSW` refer to? Is that `Haswell`?
If so, this probably shouldn't be hardcoded, and be handled via an easyblock?

**Review by boegel — CHANGES_REQUESTED** [review]

File: easybuild/easyconfigs/k/kokkos/kokkos-3.4.01-GCC-10.3.0.eb (hunk 5 above)
Comment (robert-mijakovic) [reply easybuild/easyconfigs/k/kokkos/kokkos-3.4.01-GCC-10.3.0.eb:35]: Indeed, you are right. README.md just specified that this should be used as a default option. I haven't questioned that but there is a list of supported ARCH's. The question is how to create a mapping between what archspec returns and EC.
```
KOKKOS_ARCH_OPTION(AMDAVX          HOST "AMD chip")
KOKKOS_ARCH_OPTION(ARMV80          HOST "ARMv8.0 Compatible CPU")
KOKKOS_ARCH_OPTION(ARMV81          HOST "ARMv8.1 Compatible CPU")
KOKKOS_ARCH_OPTION(ARMV8_THUNDERX  HOST "ARMv8 Cavium ThunderX CPU")
KOKKOS_ARCH_OPTION(ARMV8_THUNDERX2 HOST "ARMv8 Cavium ThunderX2 CPU")
KOKKOS_ARCH_OPTION(A64FX           HOST "ARMv8.2 with SVE Support")
KOKKOS_ARCH_OPTION(WSM             HOST "Intel Westmere CPU")
KOKKOS_ARCH_OPTION(SNB             HOST "Intel Sandy/Ivy Bridge CPUs")
KOKKOS_ARCH_OPTION(HSW             HOST "Intel Haswell CPUs")
KOKKOS_ARCH_OPTION(BDW             HOST "Intel Broadwell Xeon E-class CPUs")
KOKKOS_ARCH_OPTION(SKX             HOST "Intel Sky Lake Xeon E-class HPC CPUs (AVX512)")
KOKKOS_ARCH_OPTION(KNC             HOST "Intel Knights Corner Xeon Phi")
KOKKOS_ARCH_OPTION(KNL             HOST "Intel Knights Landing Xeon Phi")
KOKKOS_ARCH_OPTION(BGQ             HOST "IBM Blue Gene Q")
KOKKOS_ARCH_OPTION(POWER7          HOST "IBM POWER7 CPUs")
KOKKOS_ARCH_OPTION(POWER8          HOST "IBM POWER8 CPUs")
KOKKOS_ARCH_OPTION(POWER9          HOST "IBM POWER9 CPUs")
KOKKOS_ARCH_OPTION(KEPLER30        GPU  "NVIDIA Kepler generation CC 3.0")
KOKKOS_ARCH_OPTION(KEPLER32        GPU  "NVIDIA Kepler generation CC 3.2")
KOKKOS_ARCH_OPTION(KEPLER35        GPU  "NVIDIA Kepler generation CC 3.5")
KOKKOS_ARCH_OPTION(KEPLER37        GPU  "NVIDIA Kepler generation CC 3.7")
KOKKOS_ARCH_OPTION(MAXWELL50       GPU  "NVIDIA Maxwell generation CC 5.0")
KOKKOS_ARCH_OPTION(MAXWELL52       GPU  "NVIDIA Maxwell generation CC 5.2")
KOKKOS_ARCH_OPTION(MAXWELL53       GPU  "NVIDIA Maxwell generation CC 5.3")
KOKKOS_ARCH_OPTION(PASCAL60        GPU  "NVIDIA Pascal generation CC 6.0")
KOKKOS_ARCH_OPTION(PASCAL61        GPU  "NVIDIA Pascal generation CC 6.1")
KOKKOS_ARCH_OPTION(VOLTA70         GPU  "NVIDIA Volta generation CC 7.0")
KOKKOS_ARCH_OPTION(VOLTA72         GPU  "NVIDIA Volta generation CC 7.2")
KOKKOS_ARCH_OPTION(TURING75        GPU  "NVIDIA Turing generation CC 7.5")
KOKKOS_ARCH_OPTION(AMPERE80        GPU  "NVIDIA Ampere generation CC 8.0")
KOKKOS_ARCH_OPTION(AMPERE86        GPU  "NVIDIA Ampere generation CC 8.6")
KOKKOS_ARCH_OPTION(ZEN             HOST "AMD Zen architecture")
KOKKOS_ARCH_OPTION(ZEN2            HOST "AMD Zen2 architecture")
KOKKOS_ARCH_OPTION(ZEN3            HOST "AMD Zen3 architecture")
KOKKOS_ARCH_OPTION(VEGA900         GPU  "AMD GPU MI25 GFX900")
KOKKOS_ARCH_OPTION(VEGA906         GPU  "AMD GPU MI50/MI60 GFX906")
KOKKOS_ARCH_OPTION(VEGA908         GPU  "AMD GPU MI100 GFX908")
KOKKOS_ARCH_OPTION(INTEL_GEN       GPU  "Intel GPUs Gen9+")
```

File: easybuild/easyconfigs/k/kokkos/kokkos-3.4.01-GCC-10.3.0.eb (hunk 3 above)
Comment (robert-mijakovic) [reply easybuild/easyconfigs/k/kokkos/kokkos-3.4.01-GCC-10.3.0.eb:30]: Ok, resolved!

File: easybuild/easyconfigs/k/kokkos/kokkos-3.4.01-GCC-10.3.0.eb (hunk 2 above)
Comment (robert-mijakovic) [reply easybuild/easyconfigs/k/kokkos/kokkos-3.4.01-GCC-10.3.0.eb:11]: I did.

File: easybuild/easyconfigs/k/kokkos/kokkos-3.4.01-GCC-10.3.0.eb (hunk 1 above)
Comment (robert-mijakovic) [reply easybuild/easyconfigs/k/kokkos/kokkos-3.4.01-GCC-10.3.0.eb:6]: I have adapted the name.

File: easybuild/easyconfigs/k/kokkos/kokkos-3.4.01-GCC-10.3.0.eb (hunk 4 above)
Comment (robert-mijakovic) [reply easybuild/easyconfigs/k/kokkos/kokkos-3.4.01-GCC-10.3.0.eb:40]: I have added something. I don't know what a good test would be. There are only three binaries, two don't work without external stuff. I have added hpcbind -h.

File: easybuild/easyconfigs/k/kokkos/kokkos-3.4.01-GCC-10.3.0.eb (hunk 5 above)
Comment (boegel) [reply easybuild/easyconfigs/k/kokkos/kokkos-3.4.01-GCC-10.3.0.eb:35]: Picking the right value here requires a custom easyblock for Kokkos.

We have stuff like this already in the LAMMPS easyblock

Comment (ocaisa) [reply easybuild/easyconfigs/k/kokkos/kokkos-3.4.01-GCC-10.3.0.eb:35]: See https://github.com/easybuilders/easybuild-easyblocks/pull/2213/files

Comment (robert-mijakovic) [reply easybuild/easyconfigs/k/kokkos/kokkos-3.4.01-GCC-10.3.0.eb:35]: @ocaisa Is there a plan on merging this EB?

**Comment by boegel** [comment]
This would really need an easyblock, and this version of Kokkos is now long outdated, so I don't see the point in pursuing this, so closing it...

**Pull request closed without merging**